firedrake/assemble.py

```diff
@@ -41,1018 +42,987 @@
+        # FIXME This only works for 1-forms at the moment
+        key = cls._cache_key(*args, **kwargs)
+        if key is not None:
+            try:
+                return form._cache[_FORM_CACHE_KEY][key]
+            except KeyError:
+                pass
+        self = super().__new__(cls)
+        self._initialised = False
+        self.__init__(*args, **kwargs)
+        if _FORM_CACHE_KEY not in form._cache:
+            form._cache[_FORM_CACHE_KEY] = {}
+        form._cache[_FORM_CACHE_KEY][key] = self
+        return self
 
-        bcs = solving._extract_bcs(bcs)
+    @classmethod
+    @abc.abstractmethod
+    def _cache_key(cls, *args, **kwargs):
+        """Return cache key."""
 
-        self._form = form
-        self._tensor = tensor
-        self._bcs = bcs
-        self._form_compiler_params = form_compiler_parameters or {}
+    @staticmethod
+    def _skip_if_initialised(init):
+        @functools.wraps(init)
+        def wrapper(self, *args, **kwargs):
+            if not self._initialised:
+                init(self, *args, **kwargs)
+                self._initialised = True
+        return wrapper
+
+    def __init__(self, form, bcs=None, form_compiler_parameters=None):
+        super().__init__(form, bcs=bcs, form_compiler_parameters=form_compiler_parameters)
+        # Ensure mesh is 'initialised' as we could have got here without building a
+        # function space (e.g. if integrating a constant).
+        for mesh in form.ufl_domains():
+            mesh.init()
+        if any(c.dat.dtype != ScalarType for c in form.coefficients()):
+            raise ValueError("Cannot assemble a form containing coefficients where the "
+                             "dtype is not the PETSc scalar type.")
+
+
+class ParloopFormAssembler(FormAssembler):
+    """FormAssembler that uses Parloops.
+
+    Parameters
+    ----------
+    form : ufl.Form or slate.TensorBase
+        Variational form to be assembled.
+    bcs : Sequence
+        Iterable of boundary conditions to apply.
+    form_compiler_parameters : dict
+        Optional parameters to pass to the TSFC and/or Slate compilers.
+    needs_zeroing : bool
+        Should ``tensor`` be zeroed before assembling?
+
+    """
+    def __init__(self, form, bcs=None, form_compiler_parameters=None, needs_zeroing=True):
+        super().__init__(form, bcs=bcs, form_compiler_parameters=form_compiler_parameters)
         self._needs_zeroing = needs_zeroing
-        self.weight = weight
 
-    @property
-    @abc.abstractmethod
-    def result(self):
-        """The result of the assembly operation."""
+    def assemble(self, tensor=None):
+        """Assemble the form.
 
-    @property
-    @abc.abstractmethod
-    def diagonal(self):
-        """Are we assembling the diagonal of a 2-form?"""
+        Parameters
+        ----------
+        tensor : firedrake.cofunction.Cofunction or matrix.MatrixBase
+            Output tensor to contain the result of assembly.
```

Comment:
Please explain what happens if tensor is none.

Reply:
Fixed.